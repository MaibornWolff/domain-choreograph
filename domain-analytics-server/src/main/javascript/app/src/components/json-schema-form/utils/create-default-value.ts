import { JSONSchema4, JSONSchema4TypeName } from 'json-schema';

const defaultValues: Record<Exclude<JSONSchema4TypeName, 'object'>, any> = {
  string: undefined,
  number: undefined,
  integer: undefined,
  boolean: undefined,
  array: [],
  'null': null,
  any: ''
};

export function createDefaultValue(schema: JSONSchema4): any {
  const type = schema.type || 'any';
  if (type instanceof Array) {
    throw new Error('Array types are not supported');
  } else if (type === 'object') {
    return createDefaultValueForObject(schema);
  } else {
    return defaultValues[type];
  }
}

function createDefaultValueForObject(schema: JSONSchema4): Promise<object> {
  if (schema.type !== 'object') {
    throw new Error('Expect schema to be an object');
  }
  let defaultValue: any = {};
  if (schema.properties) {
    Object.entries(schema.properties).forEach(([key, property]) => {
      defaultValue[key] = createDefaultValue(property)
    });
  }
  return defaultValue;
}

