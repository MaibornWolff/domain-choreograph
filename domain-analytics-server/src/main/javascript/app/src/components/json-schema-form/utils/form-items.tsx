import { styled } from '~utils/styled';

export const FormItems = styled('div')`
    display: flex;
    flex-direction: column;

    & > * {
      margin-bottom: .5rem;
    }
`;
