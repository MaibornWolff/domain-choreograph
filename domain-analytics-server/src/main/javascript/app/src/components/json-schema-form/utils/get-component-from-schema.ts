import { FieldComponent } from 'clean-forms';
import { JSONSchema4, JSONSchema4TypeName } from 'json-schema';
import { DatePickerMsField } from '~components/json-schema-form/form-fields/date-picker-ms-field';
import { NumberTextField, TextField } from '~components/json-schema-form/form-fields/text-field';
import { JsonSchemaFormFieldProps } from '~components/json-schema-form/models/json-schema-form-field-props';

export interface JsonSchemaComponentMap {
  matcher?: { type: JSONSchema4TypeName, format?: string };
  field: FieldComponent<any, JsonSchemaFormFieldProps>;
}

const defaultMapping: JsonSchemaComponentMap[] = [
  { matcher: { type: 'integer', format: 'utc-millisec' }, field: DatePickerMsField },
  { matcher: { type: 'integer' }, field: NumberTextField },
  { matcher: { type: 'number' }, field: NumberTextField },
  { field: TextField }
];

export function getComponentFromSchema(
  schema: JSONSchema4,
  mapping: JsonSchemaComponentMap[] = defaultMapping
): FieldComponent<any, JsonSchemaFormFieldProps> {
  for (const map of mapping) {
    if (map.matcher == null) {
      return map.field;
    }
    const typeIsMatching = map.matcher.type === schema.type;
    const formatIsMatching = map.matcher.format == null || map.matcher.format === schema.format;
    if (typeIsMatching && formatIsMatching) {
      return map.field;
    }
  }
  throw new Error('Couldn\'t match the schema');
}
