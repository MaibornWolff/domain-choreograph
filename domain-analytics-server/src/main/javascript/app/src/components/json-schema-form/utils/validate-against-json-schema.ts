import { ValidationFunction } from 'clean-forms';
import { JSONSchema4 } from 'json-schema';
import { Validator } from 'jsonschema';


export function validateAgainstJsonSchema(schema: JSONSchema4): ValidationFunction {
  const validator = new Validator();
  return ({ value }) => {
    const result = validator.validate(value, schema as any);
    const errors = result.errors.map(error => {
      let message = error.message;
      let property = error.property
        .replace(/^instance\.?/, '');
      let schema = error.schema as JSONSchema4;
      if (error.name === 'required') {
        message = `is required`;
        schema = schema.properties![error.argument];
        property = joinProperties(property, error.argument);
      }

      const title = schema.title || '';
      message = `${title} ${message}`;

      return [ property, message ] as [string, string];
    });
    return errors;
  }
}

function joinProperties(path1: string, path2: string): string {
  if (path1 === '') {
    return path2;
  }
  if (path2 === '') {
    return path1;
  }
  return `${path1}.${path2}`;
}
