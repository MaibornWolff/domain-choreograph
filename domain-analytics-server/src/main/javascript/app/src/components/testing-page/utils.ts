export function escapeJavaClass(javaClass: string) {
  return javaClass.replace(/\./g, '-');
}

export function reverseJavaClassEscape(javaClass: string) {
  return javaClass.replace(/-/g, '.');
}
