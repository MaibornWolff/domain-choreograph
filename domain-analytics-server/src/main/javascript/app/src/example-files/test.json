{
  "nodes": {
    "A": {
      "id": "A",
      "name": "A",
      "scope": "root"
    }
  },
  "dependencies": [],
  "scopes": {
    "root": {
      "id": "root",
      "executionContext": "Application",
      "nodes": [
        "A"
      ]
    }
  },
  "executionContexts": {
    "Application": {
      "id": "Application",
      "scopes": [
        "root"
      ]
    }
  }
}
