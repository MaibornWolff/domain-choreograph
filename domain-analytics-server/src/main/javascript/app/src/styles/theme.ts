import darken from 'polished/lib/color/darken';
import hsl from 'polished/lib/color/hsl';
import lighten from 'polished/lib/color/lighten';
import rgba from 'polished/lib/color/rgba';

const textLight = {
  primary: 'rgba(255, 255, 255, 1)',
  secondary: 'rgba(255, 255, 255, .7)',
  disabled: 'rgba(255, 255, 255, .5)'
};

const textDark = {
  primary: 'rgba(0, 0, 0, .87)',
  secondary: 'rgba(0, 0, 0, .54)',
  disabled: 'rgba(0, 0, 0, .38)',
};

function createColorPalette(hue: number, saturation: number, light: number) {
  const difference = .035;
  return [
    hsl(hue, saturation, light + difference * 4),
    hsl(hue, saturation, light + difference * 3),
    hsl(hue, saturation, light + difference * 2),
    hsl(hue, saturation, light + difference),
    hsl(hue, saturation, light),
    hsl(hue, saturation, light - difference),
    hsl(hue, saturation, light - difference * 2),
    hsl(hue, saturation, light - difference * 3),
    hsl(hue, saturation, light - difference * 4),
  ];
}

export const mainColorIndex = 4;

const primary = createColorPalette(225, .4, .5);

const secondary =  createColorPalette(140, .5, .45);

const animations = {
  duration: {
    toggle: 300
  },
  all: `all 0.3s cubic-bezier(.25,.8,.25,1)`
};

const shadows = {
  0: '0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24)',
  1: '0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23)',
};

export const THEME = {
  type: 'light' as 'light' | 'dark',
  shadows,
  colors: {
    primary,
    secondary,
    background: {
      0: '#E0E0E0',
      1: '#F5F5F5',
      2: '#FAFAFA',
      3: '#FFFFFF',
    },
    danger: '#b73a2d',
    text: {
      light: textLight,
      dark: textDark,
      normal: textDark,
      inverse: textLight
    },
    icons: {
      light: textLight,
      dark: textDark,
      normal: textDark,
      inverse: textLight
    },
    divider: {
      normal: rgba(0, 0, 0, .12),
      inverse: rgba(255, 255, 255, .12),
    },
    border: {
      normal: rgba(0, 0, 0, .3),
    },
    graph: {
      edge: lighten(.6, 'black'),
      selectedEdge: lighten(.15, primary[mainColorIndex]),
    },
    selectedNode: primary[mainColorIndex],
  },
  jsonViewTheme: 'rjv-default',
  animations,
  dimensions: {
    borderRadius: {
      normal: '3px',
      big: '5px',
    },
    border: {
      normal: '1px',
      thick: '2px'
    }
  }
};

export type Theme = typeof THEME;

const DARK_THEME: Theme = {
  type: 'dark',
  shadows,
  colors: {
    primary,
    secondary,
    background: {
      0: '#191919',
      1: '#343434',
      2: '#424242',
      3: '#303030',
    },
    text: {
      light: textLight,
      dark: textDark,
      normal: textLight,
      inverse: textDark,
    },
    icons: {
      light: textLight,
      dark: textDark,
      normal: textLight,
      inverse: textDark,
    },
    danger: '#d65246',
    divider: {
      normal: rgba(255, 255, 255, .12),
      inverse: rgba(0, 0, 0, .12),
    },
    border: {
      normal: rgba(255, 255, 255, .3),
    },
    graph: {
      edge: darken(.5, 'white'),
      selectedEdge: primary[mainColorIndex],
    },
    selectedNode: primary[mainColorIndex],
  },
  animations,
  jsonViewTheme: 'bright',
  dimensions: THEME.dimensions
};

export const THEMES = {
  light: THEME,
  dark: DARK_THEME
};
export type ThemeKey = keyof typeof THEMES;
