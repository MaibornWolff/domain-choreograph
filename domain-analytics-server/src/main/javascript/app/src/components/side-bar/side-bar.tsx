import { ClassNames, css } from '@emotion/core';
import * as React from 'react';
import ReactJson from 'react-json-view';
import { Markdown } from '~components/markdown/markdown';
import { SectionContainer } from '~containers/section.container';
import { Theme } from '~styles/theme';
import { ChoreoNode } from '~types/choreo-graph';
import { styled } from '~utils/styled';
import { Resizable } from '../resizable/resizable';

const StyledHeader = styled('h2')`
    font-size: 1.3rem;
    font-weight: bold;
    overflow: hidden;
    text-overflow: ellipsis;
    margin: 0 0 .8rem 0;
`;

const resizableWrapperStyle = css`
  z-index: 3;
  position: absolute;
  right: 0;
  top: 0;
  bottom: 0;
`;

const StyledResizable = styled(Resizable)`
  padding: .5rem 1rem;
  overflow-x: auto;
  overflow-y: scroll;
  box-sizing: border-box;
  background: ${({ theme }) => theme.colors.background[2]};
  box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);
`;

const NoSelection = styled('div')`
  display: flex;
  justify-content: center;
  align-items: center;
  color: ${({ theme }) => theme.colors.text.normal.secondary};
  font-size: 1.3rem;
  height: 100%;
`;

const JsonWrapper = styled('div')`
    font-size: .8rem;

    .react-json-view {
      background: none !important;
    }
`;

const Danger = styled('div')`
  font-size: .8rem;
  color: ${({ theme }) => theme.colors.danger};
  margin-bottom: 1rem;
`;

const PreviewLabel = styled('pre')`
  color: ${({ theme }) => theme.colors.text.dark.secondary};
  font-size: .8rem;
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0 0 1rem;
  white-space: pre-wrap;
`;

const ExceptionLabel = styled('div')`
    font-weight: bold;
`;

const ExceptionMessage = styled('pre')`
    white-space: normal;
    margin: .5rem 0;
`;

export interface SideBarProps {
  node: ChoreoNode | null;
  theme: Theme;
}

export interface SideBarState {
}

export class SideBar extends React.Component<SideBarProps, SideBarState> {

  public render() {
    const { node, theme } = this.props;
    if (!node) {
      return null;
    }

    return (
      <ClassNames>
        {({ css }) => (
          <StyledResizable wrapperClassName={css(resizableWrapperStyle)}>
            {
              node
                ? (
                  <div>
                    <StyledHeader>{node.name}</StyledHeader>
                    {node.preview && <PreviewLabel>{node.preview}</PreviewLabel>}
                    {node.exception
                      ? (
                        <SectionContainer header={'Exception'} isInitiallyOpen={true}>
                          <Danger>
                            <ExceptionLabel>Exception:</ExceptionLabel>
                            <ExceptionMessage>{node.exception.detailMessage}</ExceptionMessage>
                          </Danger>
                        </SectionContainer>
                      )
                      : null
                    }
                    {node.doc
                      ? <SectionContainer header={'Documentation'} isInitiallyOpen={true}><Markdown
                        markdown={node.doc}/></SectionContainer>
                      : null}
                    <SectionContainer header={'Inspection'} isInitiallyOpen={true}>
                      <JsonWrapper>
                        <ReactJson
                          src={getNodeValueAsJson(node)}
                          theme={theme.jsonViewTheme}
                          name={getNodeNameForJson(node)}
                          indentWidth={2}
                          collapsed={1}
                        />
                      </JsonWrapper>
                    </SectionContainer>
                  </div>
                )
                : <NoSelection>No Selection</NoSelection>
            }
          </StyledResizable>
        )}
      </ClassNames>
    );
  }
}

function getNodeValueAsJson(node: ChoreoNode): object {
  switch (typeof node.value) {
    case 'object':
      return node.value as object;
    default:
      return { [node.name]: node.value };
  }
}

function getNodeNameForJson(node: ChoreoNode): string | null {
  switch (typeof node.value) {
    case 'object':
      return node.name;
    default:
      return null;
  }
}
