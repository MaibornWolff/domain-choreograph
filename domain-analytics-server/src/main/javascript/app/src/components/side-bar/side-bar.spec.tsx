import React from 'react';
import { THEME } from '~styles/theme';
import { renderWithProviders } from '~utils/test.utils';
import { SideBar } from './side-bar';

describe('SideBar', () => {
  it('should render', () => {
    expect(renderWithProviders(<SideBar theme={THEME} node={{
      id: 'id',
      name: 'name',
      scope: 'scope',
      value: { test: '123' },
      preview: 'preview'
    }}/>)).toMatchSnapshot();
  });
});
