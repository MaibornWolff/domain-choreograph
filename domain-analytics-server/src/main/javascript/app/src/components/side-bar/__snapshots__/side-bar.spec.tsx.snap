// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar should render 1`] = `
.c8 {
  height: 24px;
  width: 24px;
}

.c8 svg {
  fill: rgba(0,0,0,.87);
}

.c7 {
  -webkit-transition: -webkit-transform 300ms ease;
  -webkit-transition: transform 300ms ease;
  transition: transform 300ms ease;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: .5rem 0;
  cursor: pointer;
}

.c10 {
  font-size: 1rem;
  padding: 0;
  margin: 0;
  font-weight: bold;
}

.c11 {
  height: 1px;
  background: rgba(0,0,0,.38);
  -webkit-flex: 1 1 0;
  -ms-flex: 1 1 0;
  flex: 1 1 0;
  margin-left: .5rem;
}

.c9 {
  height: 1px;
  background: rgba(0,0,0,.38);
  -webkit-flex: 0 0 1rem;
  -ms-flex: 0 0 1rem;
  flex: 0 0 1rem;
  margin-right: .5rem;
}

.c12 {
  overflow: hidden;
  -webkit-transition: height 300ms ease;
  transition: height 300ms ease;
  height: auto;
}

.c13 {
  box-sizing: border-box;
  padding: .5rem 0 .5rem 1rem;
}

.c0 {
  position: absolute;
  right: 0;
  top: 0;
  bottom: 0;
  height: 100%;
}

.c3 {
  overflow: auto;
  height: 100%;
}

.c1 {
  position: absolute;
  width: 50px;
  left: -25px;
  top: 0;
  bottom: 0;
  cursor: col-resize;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4 {
  font-size: 1.3rem;
  font-weight: bold;
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0 0 .8rem 0;
}

.c2 {
  padding: .5rem 1rem;
  overflow-x: auto;
  overflow-y: scroll;
  box-sizing: border-box;
  background: #FAFAFA;
  box-shadow: 0 1px 3px rgba(0,0,0,0.12),0 1px 2px rgba(0,0,0,0.24);
}

.c2 .resizable-wrapper {
  z-index: 3;
  position: absolute;
  right: 0;
  top: 0;
  bottom: 0;
}

.c14 {
  font-size: .8rem;
}

.c14 .react-json-view {
  background: none !important;
}

.c5 {
  color: rgba(0,0,0,.54);
  font-size: .8rem;
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0 0 1rem;
  white-space: pre-wrap;
}

<div
  class=".resizable-wrapper c0"
  style="width:350px"
>
  <div
    class="c1"
  />
  <div
    class="c2 c3"
  >
    <div>
      <h2
        class="c4"
      >
        name
      </h2>
      <pre
        class="c5"
      >
        preview
      </pre>
      <div>
        <div
          class="c6"
        >
          <div
            class="c7 c8"
          >
            [object Object]
          </div>
          <div
            class="c9"
          />
          <h2
            class="c10"
          >
            Inspection
          </h2>
          <div
            class="c11"
          />
        </div>
        <div
          class="c12"
        >
          <div
            class="c13"
          >
            <div
              class="c14"
            >
              <div
                class="react-json-view"
                style="font-family:monospace;cursor:default;background-color:rgba(0, 0, 0, 0);position:relative"
              >
                <div
                  class="pretty-json-container object-container"
                >
                  <div
                    class="object-content"
                  >
                    <div
                      class="object-key-val"
                    >
                      <span>
                        <span
                          style="display:inline-block;cursor:pointer"
                        >
                          <div
                            class="icon-container"
                            style="display:inline-block;width:17px"
                          >
                            <span
                              class="expanded-icon"
                            >
                              <svg
                                fill="currentColor"
                                style="vertical-align:top;color:#586e75;height:1em;width:1em;padding-left:2px"
                                viewBox="0 0 15 15"
                              >
                                <path
                                  d="M0 5l6 6 6-6z"
                                />
                              </svg>
                            </span>
                          </div>
                          <span
                            style="display:inline-block;color:#002b36;letter-spacing:0.5px;font-style:none;vertical-align:top;opacity:0.85;:hover:[object Object]"
                          >
                            <span
                              class="object-key"
                            >
                              <span
                                style="vertical-align:top"
                              >
                                "
                              </span>
                              <span>
                                name
                              </span>
                              <span
                                style="vertical-align:top"
                              >
                                "
                              </span>
                            </span>
                            <span
                              style="display:inline-block;margin:0px 5px;color:#002b36;vertical-align:top"
                            >
                              :
                            </span>
                          </span>
                          <span
                            style="display:inline-block;cursor:pointer;font-weight:bold;color:#002b36"
                          >
                            {
                          </span>
                        </span>
                        <div
                          class="object-meta-data"
                          style="display:inline-block;padding:0px 0px 0px 10px"
                        >
                          <span
                            class="object-size"
                            style="color:rgba(0, 0, 0, 0.3);border-radius:3px;font-style:italic;margin:0px 6px 0px 0px;cursor:default"
                          >
                            1 item
                          </span>
                          <span
                            class="copy-to-clipboard-container"
                          >
                            <span
                              style="cursor:pointer;display:inline"
                            >
                              <span
                                class="copy-icon"
                              >
                                <svg
                                  fill="currentColor"
                                  preserveAspectRatio="xMidYMid meet"
                                  style="vertical-align:top;color:#268bd2;font-size:15px;margin-right:3px;height:1em;width:1em"
                                  viewBox="0 0 40 40"
                                >
                                  <g>
                                    <path
                                      d="m30 35h-25v-22.5h25v7.5h2.5v-12.5c0-1.4-1.1-2.5-2.5-2.5h-7.5c0-2.8-2.2-5-5-5s-5 2.2-5 5h-7.5c-1.4 0-2.5 1.1-2.5 2.5v27.5c0 1.4 1.1 2.5 2.5 2.5h25c1.4 0 2.5-1.1 2.5-2.5v-5h-2.5v5z m-20-27.5h2.5s2.5-1.1 2.5-2.5 1.1-2.5 2.5-2.5 2.5 1.1 2.5 2.5 1.3 2.5 2.5 2.5h2.5s2.5 1.1 2.5 2.5h-20c0-1.5 1.1-2.5 2.5-2.5z m-2.5 20h5v-2.5h-5v2.5z m17.5-5v-5l-10 7.5 10 7.5v-5h12.5v-5h-12.5z m-17.5 10h7.5v-2.5h-7.5v2.5z m12.5-17.5h-12.5v2.5h12.5v-2.5z m-7.5 5h-5v2.5h5v-2.5z"
                                    />
                                  </g>
                                </svg>
                              </span>
                            </span>
                          </span>
                        </div>
                      </span>
                      <div
                        class="pushed-content object-container"
                      >
                        <div
                          class="object-content"
                          style="margin-left:6px"
                        >
                          <div
                            class="variable-row"
                            style="padding-top:3px;padding-right:5px;padding-bottom:3px;border-left:1px solid rgb(235, 235, 235);:hover:[object Object];padding-left:10px"
                          >
                            <span>
                              <span
                                class="object-key"
                                style="display:inline-block;color:#002b36;letter-spacing:0.5px;font-style:none;vertical-align:top;opacity:0.85;:hover:[object Object]"
                              >
                                <span
                                  style="vertical-align:top"
                                >
                                  "
                                </span>
                                <span
                                  style="display:inline-block"
                                >
                                  test
                                </span>
                                <span
                                  style="vertical-align:top"
                                >
                                  "
                                </span>
                              </span>
                              <span
                                style="display:inline-block;margin:0px 5px;color:#002b36;vertical-align:top"
                              >
                                :
                              </span>
                            </span>
                            <div
                              class="variable-value"
                              style="display:inline-block;padding-right:6px;position:relative;cursor:default"
                            >
                              <div
                                style="display:inline-block;color:#cb4b16"
                              >
                                <span
                                  class="data-type-label"
                                  style="font-size:11px;margin-right:4px;opacity:0.8"
                                >
                                  string
                                </span>
                                <span
                                  class="string-value"
                                  style="cursor:default"
                                >
                                  "123"
                                </span>
                              </div>
                            </div>
                            <span
                              class="copy-to-clipboard-container"
                            >
                              <span
                                style="cursor:pointer;display:inline"
                              >
                                <span
                                  class="copy-icon"
                                >
                                  <svg
                                    fill="currentColor"
                                    preserveAspectRatio="xMidYMid meet"
                                    style="vertical-align:top;color:#268bd2;font-size:15px;margin-right:3px;height:1em;width:1em"
                                    viewBox="0 0 40 40"
                                  >
                                    <g>
                                      <path
                                        d="m30 35h-25v-22.5h25v7.5h2.5v-12.5c0-1.4-1.1-2.5-2.5-2.5h-7.5c0-2.8-2.2-5-5-5s-5 2.2-5 5h-7.5c-1.4 0-2.5 1.1-2.5 2.5v27.5c0 1.4 1.1 2.5 2.5 2.5h25c1.4 0 2.5-1.1 2.5-2.5v-5h-2.5v5z m-20-27.5h2.5s2.5-1.1 2.5-2.5 1.1-2.5 2.5-2.5 2.5 1.1 2.5 2.5 1.3 2.5 2.5 2.5h2.5s2.5 1.1 2.5 2.5h-20c0-1.5 1.1-2.5 2.5-2.5z m-2.5 20h5v-2.5h-5v2.5z m17.5-5v-5l-10 7.5 10 7.5v-5h12.5v-5h-12.5z m-17.5 10h7.5v-2.5h-7.5v2.5z m12.5-17.5h-12.5v2.5h12.5v-2.5z m-7.5 5h-5v2.5h5v-2.5z"
                                      />
                                    </g>
                                  </svg>
                                </span>
                              </span>
                            </span>
                          </div>
                        </div>
                      </div>
                      <span
                        class="brace-row"
                      >
                        <span
                          style="display:inline-block;cursor:pointer;font-weight:bold;color:#002b36;padding-left:3px"
                        >
                          }
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
